text/0000-constants-in-patterns.md

```diff
@@ -0,0 +1,302 @@
+- **Fallback to `==`.**
+  When the constant fails the structural equality test, instead of rejecting the code outright, we could accept it and compare with `==` instead.
+  This might be surprising since the user did not ask for `==` to be invoked.
+  This also makes it harder to later add support for libraries controlling matching themselves, since one can already match on everything that has `PartialEq`,
+  but that latter point could be mitigated by only allowing such matching when a marker trait is implemented.
+  There currently does not seem to be sufficient motivation for doing this, and the RFC as proposed is forward-compatible with doing this in the future should the need come up.
+- Do something that violates the core design principles laid out in the "Motivation" section.
+  This is considered a no-go by this RFC:
+  having possible behavior changes upon "outlining" a fieldless enum variant into a constant is too big of a footgun,
+  and allowing matching without opt-in from the crate that defines the type makes abstraction and semver too tricky to maintain (consider that private fields would be compared when matching).
+
+# Prior art
+[prior-art]: #prior-art
+
+[RFC 1445][rfc-1445] defines basically the same checks as this RFC; this RFC merely spells them out more clearly for cases the old RFC did not explicitly over (nested user-defined types, raw pointers),
+and adjusts to decisions that have been made in the mean time (accepting floating-point patterns).
+
+This RFC came out of discussions in a [t-lang design meeting](https://github.com/rust-lang/lang-team/issues/220).
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None at this time.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- At some point, we might want to stabilize the `StructuralPartialEq` trait.
+  There are however plenty of open questions here:
+  + Should `StructuralPartialEq` be an unsafe trait?
+    That trait has a clear semantic meaning, so making it `unsafe` to be able to rely on it for soundness is appealing.
+    In particular, we could then be *sure* that pattern matching always has the same semantics as `==`.
+    However, that would mean `forbid(unsafe_code)` crates could not `derive(PartialEq)` any more, which is clearly not acceptable.
+    As a consequence of `StructuralPartialEq` being safe, it is actually possible to write patterns that behave different from the `PartialEq` that is explicitly defined and intentionally exposed on that type,
+    but only if one of the involved crates implements `StructuralPartialEq` incorrectly.
+    This can lead to semver issues and logic bugs, but that is all allowed for safe traits.
+    However, this also means unsafe code cannot rely on `==` and pattern matching having the same behavior.
+    One could consider doing some special magic to `derive(PartialEq)` to remain compatible with `forbid(unsafe_code)`, but that still seems suboptimal.
```

Comment:
Such magic already exists, it's called `#[allow_internal_unsafe]`.
It is currently used on definitions of `format_args!` and `thread_local!`.

Reply:
Oh I didn't know about that -- I'll mention it in the RFC, thanks.